students/EricAdams/session04/mailroom.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+
+"""
+Used Chris's solution to session 3 version and added to it to create this,
+(session 4) version
+
+Mailroom Exercise -- as of Session 4--  dictionaries yes  Exceptions no
+
+Note: this is not the most robust or flexible code -- but it does the
+job with the simplest of Python data structures.
+
+Adding dictionaries
+Adding function send_thank_you_to_file.
+"""
+
+from textwrap import dedent  # nifty utility!
+import sys
+import math
+
+# In memory representation of the donor database
+# using a tuple for each donor
+# -- kind of like a record in a database table
+# the donations are in a list -- so you can add to them
+# Note the mutable inside an immutable .. that is inside a mutable!
+
+
+# Making this a global, so it can be accessed from various functions
+donor_db = {"William Gates, III": [653772.32, 12.17],
+            "Jeff Bezos": [877.33],
+            "Paul Allen": [663.23, 43.87, 1.32],
+            "Mark Zuckerberg": [1663.23, 4300.87, 10432.0]}
+
+
+# loop through the donor list and print the 0th element of the list
+def print_donors():
+    """
+    loop through the donor list and print the 0th element of the list
+    which is the donors name
+    donor name is now the key, since changing over to dict
+    """
+
+    print("Donor list:\n")
+    for donor in donor_db.keys():
+        print(donor)
+
+
+def find_donor(name):
+    """
+    Find a donor in the donor db
+
+    :param: the name of the donor
+
+    :returns: The donor data structure -- None if not in the donor_db
+    """
+    donor = name.lower()
+    for key in donor_db.keys():
```

Comment:
so this kinda kills the point of using a dict. The great thing about a dict is fast lookup. But you're right, it needs to be an exact match. so you do need to do this loop if you maynot have an exact match.

Another solution -- use a "standardized" version of the name as the dict key (maybe lower(), spaces removed, or?

then store the "proper" version of the name in the values.

that will give you the quick lookup.